### PR TITLE
Implement new REST API to retrieve API artifacts

### PR DESCRIPTION
**Pull request by Jayanie**
## Purpose

- This PR fixes wso2-enterprise/choreo#5941
- Introduces a New REST API as /retrieve-api-artifacts with a post resource to pass the list of the API UUIDs as a payload in the request and obtain the API artifacts zip.

- This PR ports the fix done in https://github.com/wso2/carbon-apimgt/pull/10636

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/APIArtifactGeneratorUtil.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
+package org.wso2.carbon.apimgt.impl.gatewayartifactsynchronizer;
```
Comment (mpmunasinghe) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/APIArtifactGeneratorUtil.java:1]: Missing license header

Comment (Jayanie) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/APIArtifactGeneratorUtil.java:18]: Fixed with 08544ab05fc0c99881d8b53871eeab8c0fe5d14e

File: components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/resources/api.yaml [hunk 2]
```diff
@@ -476,6 +476,38 @@ paths:
           description: Unexpected error
           schema:
             $ref: '#/definitions/Error'
+  /retrieve-api-artifacts:
+    post:
+      summary: Get API runtime artifacts from ID list
+      description: |
+        This will provide access to API runtime artifacts in database for a given API UUID list.
+      parameters:
+        - $ref: '#/parameters/requestedTenant'
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/resources/api.yaml:485]: - $ref: '#/parameters/requestedOrg'

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java [hunk 3]
```diff
@@ -384,11 +384,86 @@ public void deleteGatewayArtifacts(String apiId) throws APIManagementException {
...
+        int apiIdsChunkSize = SQLConstants.API_ID_CHUNK_SIZE;
+        while (apiIdsChunkSize < apiIdListSize) {
+            apiIdsChunk.add(apiIds.subList(apiIdArrayIndex, apiIdsChunkSize));
+            apiIdListSize = apiIdListSize - SQLConstants.API_ID_CHUNK_SIZE;
+            apiIdArrayIndex = apiIdArrayIndex + SQLConstants.API_ID_CHUNK_SIZE;
+            apiIdsChunkSize = apiIdsChunkSize + SQLConstants.API_ID_CHUNK_SIZE;
+        }
+        if (apiIdListSize > 0) {
+            apiIdsChunk.add(apiIds.subList(apiIdArrayIndex, apiIdArrayIndex + apiIdListSize));
+        }
```
Comment (vlgunarathne) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java:403]: Please check if you can simplify the logic here as the following example.

```java
        List<String> apiIds = Arrays.asList("id1","id2","id3","id4","id5","id6","id7");
        int PARTITION_SIZE = 3;
        int INDEX = 0;
        int ID_COUNT = apiIds.size();
        List<List<String>> apiPartitions = new ArrayList<>();
        while (INDEX < ID_COUNT) {
            apiPartitions.add(apiIds.subList(INDEX, Math.min(INDEX + PARTITION_SIZE, ID_COUNT)));
            INDEX += PARTITION_SIZE;
        }
        System.out.println(apiPartitions);
```

Output:
```
[[id1, id2, id3], [id4, id5, id6], [id7]]
```

**Review by malinthaprasan — APPROVED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java [hunk 4]
```diff
@@ -384,11 +384,86 @@ public void deleteGatewayArtifacts(String apiId) throws APIManagementException {
...
+        List<APIRuntimeArtifactDto> apiRuntimeArtifactDtoList = new ArrayList<>();
+
+        for (List<String> apiIdList: apiIdsChunk) {
+            apiRuntimeArtifactDtoList = new ArrayList<>(apiRuntimeArtifactDtoList);
```
Comment (vlgunarathne) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java:407]: Is this line required? It seems the same `apiRuntimeArtifactDtoList` is getting initialized.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java (hunk 3 above)
Comment (Jayanie) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java:398]: Fixed with b47708f01abb761d5651b988a460491bb59d2e9d

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java (hunk 4 above)
Comment (Jayanie) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java]: Fixed with b47708f01abb761d5651b988a460491bb59d2e9d

**Review by vlgunarathne — APPROVED** [review]

**Review by mpmunasinghe — APPROVED** [review]
LGTM

**Pull request merged**